### ISSUE TITLE
How to let rtags support cuda (nvcc *.cu) indexing?

### ISSUE THREAD

**Issue by mraway** [Andersbakken/rtags#1083]
### Please mark appropriate
 - [x] rtags (rdm/rc/rp)
 - Emacs Lisp
   - [ ] ac-rtags.el
   - [ ] company-rtags.el
   - [ ] helm-rtags.el
   - [ ] flycheck-rtags.el
   - [ ] ivy-rtags.el
   - [ ] rtags.el

### Problem description
Describe briefly the problem.
I wish rtags could find ways to support cuda indexing better.
Take this repo as an example:
https://github.com/tbennun/cudnn-training

Run
cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1 .
gives no actual cuda compiling command:
{
  "directory": "/private/home/wzh/cudnn-training",
  "command": "/private/home/wzh/sbin/c++   -I/include -I/public/apps/cuda/8.0/include   -std=c++11   -o CMakeFiles/trainlenet.dir/readubyte.cpp.o -c /private/home/wzh/cudnn-training/readubyte.cpp",
  "file": "/private/home/wzh/cudnn-training/readubyte.cpp"
}

Only the c++ files get indexed. But cuda (*.cu) files are c++ files, I wish there are better ways to support .cu in cmake (or a workaround).

Thanks!

#### Expected behavior
Describe what you expected to happen.

#### Actual behavior
Describe what actually happened.

### Environment
- Your operating system:

- LLVM/Clang version:

### Feature request

Describe the feature we should add to **RTags**.

**Comment by nabajour** [comment]
I got the same issue when trying to make a compile database for clang-tidy, did you find any solution? I think it's a CMake more than an rtags issue.

**Comment by BenByington** [comment]
I'm unsure if the previous commenters are using cmake 3.8 or higher, with CUDA as an officially supported language, or using the older `FindCUDA` module, but personally when I have cmake export compile commands my `.cu` files are included just as desired. 

However I think this ticket is still very relevant, as those entries are ignored upon importing into `rtags`.  My `.cu` and `.cuh` files are reported as not indexed, and if I tell `rc` to dump out its compile commands, the entries for files with those extensions are missing despite being present in the originally imported json.

I tried briefly to muck about in the source code to see if I could hack in a workaround.  My understanding is that clang understands cuda and should be able to handle these files.  Even if `rtags` had to treat them like vanilla C++ files and tripped up a bit over keywords like `__device__` or kernel launching syntax, it would still be very useful.  I did find a couple places with hard coded lists of expected file extensions, and had hoped by adding `.cu` and `.cuh` to those lists I could trick `rtags` into at least trying to process these files, but that did not seem to affect anything.

I'm more than happy to do the work for this change if anyone can point me towards where the missing pieces may lie and some hints as to how the bigger pieces fit together.

**Comment by BenByington** [comment]
I actually hunkered down and spent a few hours on this.  I found some tweaks that get me a lot of the way there, but something is still missing.

To be specific the changes were:

- guessLanguageFromSourceFile: I made it return `CPlusPlus` when a "cu" suffix is found
- Source::Parse: When handling the `-x` flag I added detection for a `cu` argument.  There I both set the language to `CPlusPlus`, as well as change the argument to `cuda`, because for whatever reason clang differs from `nvcc` and expects `-x cuda` (`-x cu` makes it have a hard error)
- in the rtc project changed `Path::IsSource` and `Path::IsHeader` to respect `.cu` and `.cuh` respectively.

With those changes `rtags` does seem to try and parse the cuda entries in the `compile_commands.json`.  The `rdm` logs show it being processed, and despite some "handleInclude failed" errors related to some clang specific cuda wrappers, it otherwise doesn't complain.  However things don't seem to be fully indexed and as I mentioned I think something is still missing.

In particular, I happened to notice that if my cuda compile command does not specify `--std=c++<xx>` things are almost fully parsed, and I can successfully jump around between declarations/definition/usages/etc even within `.cu` files.  However I get a lot of diagnostic messages about how some code I'm using is a "c++11 extension".  That's fine, I guess clang is just assuming a different default than my usual compiler.  However adding `--std=c++<xx>` seems to break indexing in cuda files.  

I tried with both c++11 and c++14, and the former seemed to work while the latter seemed to completely confuse rtags.  `rdm` still did not complain about anything extra when handling cuda files, but none of the code navigation commands worked when jumping into or around within cuda files.  I would have thought this more likely a clang problem than an rtags problem, except I happened to notice I did my work on a relatively older version of `master`.  It was probably only 2-3 weeks old, but once I updated to the latest and applied my changes again, now both c++11 and c++14 broke indexing.  Since the behavior changed when all that changed was the rtags binary I suspect I'm still missing something in my tweaks, though at this stage I'm not sure what. 

I'm happy to either keep bashing away on this (time allowing), or supply a code delta and a toy cuda project that should serve as a test case.

**Comment by BenByington** [comment]
I have to return to my regularly scheduled workload soon, but one more quick update first.

I upgraded from clang 6 to clang 9, since I'm using a brand spanking new cuda toolkit clang 6 certainly never knew about, and now the `handleInclude failed` errors I alluded to in my last post have resolved.  Indexing seems to only partially work still (though it's more functional than I realized yesterday.  The set of symbols I was trying to use as a test for jumping between declaration/definition/usage was too small).  Some of the failures are on what I would assume are simple cases though, and I did see behaviour changes both when changing the language version as well as the rtags version, so I feel I should attempt understand that somewhat before even thinking of issuing a PR here to incorporate my changes.  

If you've any tips for  debugging indexing failures they are most welcome.  I've not thought about it deeply yet but my first thought is that I'm going to have to try and root around in the raw AST received from clang and the resulting database rtags accumulates, which sounds prohibitive...
